채팅화면 제작

blackSooooo opened the pull request.

comforest commented on package.json:
```diff
@@ -0,0 +1,57 @@
+    "mini-css-extract-plugin": "1.4.1",
+    "node-sass": "5.0.0",
+    "react": "17.0.2",
+    "react-dom": "17.0.2",
```
우리 프로젝트에서는 react-route-dom 보다는 next js 를 이용할 예정입니다.
다다음 과제 쯤에서 공부하실 것 같아요 ㅎㅎ

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+const ChatPage = ({location, history}) => {
+    const name = location.state;
+    const [message, setMessage] = useState([]);
+    const userText=['안녕하세요', '샘플 데이터입니다', '반가워요', '날씨가 어때요?', '감사합니다', '좋은하루 되세요', '사랑해요ㅎ.ㅎ'];
```
```suggestion
    const userTexts=['안녕하세요', '샘플 데이터입니다', '반가워요', '날씨가 어때요?', '감사합니다', '좋은하루 되세요', '사랑해요ㅎ.ㅎ'];
```
OR
```suggestion    
    const userTextList=['안녕하세요', '샘플 데이터입니다', '반가워요', '날씨가 어때요?', '감사합니다', '좋은하루 되세요', '사랑해요ㅎ.ㅎ'];
```
userText 라고 변수명을 지으면 느낌이 String 변수의 느낌이 나네요.
실상은 List 인데 말이죠
보통 복수형, List 등을 붙입니다.

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+    const name = location.state;
+    const [message, setMessage] = useState([]);
+    const userText=['안녕하세요', '샘플 데이터입니다', '반가워요', '날씨가 어때요?', '감사합니다', '좋은하루 되세요', '사랑해요ㅎ.ㅎ'];
+    const [value, setValue] = useState('');
```
```suggestion
    const [inputMessage, setInputMessage] = useState('');
```
변수명을 value 라고 이름을 지으시면 어떤 값을 담고 있는지 전혀 유추를 할 수가 없습니다.
자료형까지 말이죠.

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+    const [message, setMessage] = useState([]);
+    const userText=['안녕하세요', '샘플 데이터입니다', '반가워요', '날씨가 어때요?', '감사합니다', '좋은하루 되세요', '사랑해요ㅎ.ㅎ'];
+    const [value, setValue] = useState('');
+    const onChange = (event) => {
```
```suggestion
    const onChangeInputMessage = (event) => {
```

변수/함수 네이밍 고민을 해보시는게 좋을 것 같아요.
나중에 코드가 무척 길어지면... onChange가 무엇이 변경되었을 때 호출 되는건지 알기 위해선 많이 피곤해집니다^^

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+        setValue(event.target.value);
+    };
+    
+    const sendFunc = () => {
```
```suggestion
    const sendMessage = () => {
```

comforest commented on src/pages/MainPage.js:
```diff
@@ -0,0 +1,22 @@
+
+const ChatPage = ({ location, history }) => {
+    const [value, setValue] = useState('');
+    const onChange = (event) => {
```
```suggestion
    const onChangeName = (event) => {
```

comforest commented on src/pages/MainPage.js:
```diff
@@ -0,0 +1,22 @@
+import { Link } from 'react-router-dom';
+
+const ChatPage = ({ location, history }) => {
+    const [value, setValue] = useState('');
```
```suggestion
    const [name, setName] = useState('');
```

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+
+    return (
+        <div className="container">
+            <div className="header">
+                {name}
+            </div>
+            <div className="chatBox">
+                {message.map(item => {
+                    return item['isMine'] ? <div style={{textAlign: 'right'}}><div className="mySendBox">{item['text']}</div></div> : <div className="userSendBox">{item['text']}</div>
+                })}
+            </div>
+            <div className="inputBox">
+                <textarea type="text" value={value} onChange={onChange} className="textBox"/>
+                <button className={value.length === 0? "sendBox" : "sendTextingBox"} onClick={sendFunc}>전송</button>
+            </div>
```
확장성(추후 수정/추가 사항)을 고려했을 땐 각각 컴포넌트로 만들어주는 것도 좋은 생각 일 수 있습니다.
특히 36번줄은 많이 복잡해 보이네요.

comforest commented on src/index.js:
```diff
@@ -0,0 +1,17 @@
+// If you want to start measuring performance in your app, pass a function
+// to log results (for example: reportWebVitals(console.log))
+// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
+reportWebVitals();
```
처음 보는거라 좀 살펴봤더니 괜찮은 라이브러리 같아요.
그런데 이걸 사용하신 의도가 궁금하네요.
코멘트로 답변 달아주세요 ㅎㅎ

comforest commented on src/App.test.js:
```diff
@@ -0,0 +1,8 @@
+import { render, screen } from '@testing-library/react';
+import App from './App';
+
+test('renders learn react link', () => {
+  render(<App />);
+  const linkElement = screen.getByText(/learn react/i);
+  expect(linkElement).toBeInTheDocument();
+});
```
Test 코드는 알고 짜신건가요?
아님 init 할때 기본적으로 제공 되는건가요?

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+                })}
+            </div>
+            <div className="inputBox">
+                <textarea type="text" value={value} onChange={onChange} className="textBox"/>
```
input 이 아니라 textarea를 사용하신 이유가 있을까요?
엔터도 전송이 아니라 new line 처리 되서 오히려 불편하고
textarea 크기도 멋대로 변경 할 수 있고, 이 때문인지 모르겠지만 하단 css 구성도 이쁘게 되어 있지 않네요

blackSooooo replied:
저는 PC카톡 생각하면서 만들어서 입력할 때 여러줄 가능하게 하려고 textarea를 사용했어요,, textarea 크기를 멋대로 변경할 수 있는지는 몰랐네요. 알겠습니다!

comforest commented on src/pages/MainPage.js:
```diff
@@ -0,0 +1,22 @@
+        <div>
+            <h1>대화 상대를 입력해주세요</h1>
+            <div>
+              <input type="text" placeholder="이름" value={value} onChange={onChange} />
```
Enter 키를 눌렀을 때 버튼 클릭 처리가 되면 좋을 것 같아요!

tankboy091e replied:
참고로 키보드 리스너를 붙이지 않고 Form의 onSubmit을 이용해서 구현해 보실 수 있습니다.

blackSooooo replied:
말씀해주신대로 form의 onSubmit 과 useRef를 이용해서 결과값만 바꾸고 페이지 이동을 하려 했는데 name에 못받아와서 혹시 어떤식으로 코드 작성을 할 수 있을까요?
```
import React, { useState, useRef } from 'react';
import { Link } from 'react-router-dom';

const ChatPage = ({ location, history }) => {
    const [name, setName] = useState('');
    const nameRef = useRef();
    const startChat = () => {
      setName(nameRef.current.value); 
    }
    return (
        <div>
            <h1>대화 상대를 입력해주세요{name}</h1>
            <div>
              <form onSubmit={startChat}>
                <input type="text" placeholder="이름" ref={nameRef}/>
                <Link to={{pathname: "/chat", state: name}}>
                  <input type="submit" value="채팅방 입장"/>
                </Link>
              </form>
            </div>
        </div>
    )
}
```


이렇게 작성해봤는데 name을 못받은 채로 넘어가버리네요,,

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+        if(value === '')  return;
+        let tok = false;
+        let idx = 0;
+        if(Math.floor(Math.random() * 2)){
```
```suggestion
        if(Math.random() < 0.5){
```
50% 확률이면 이게 더 깔끔한 것 같아요

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+
+const ChatPage = ({location, history}) => {
+    const name = location.state;
+    const [message, setMessage] = useState([]);
```
```suggestion
    const [messageList, setMessageList] = useState([]);
```

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+    
+    const sendFunc = () => {
+        if(value === '')  return;
+        let tok = false;
```
약어는 추후 코드 읽는데 매우 방해되는 애들 중 하나입니다...
변수가 무엇을 뜻하는지 모르기 때문이죠
작성자 본인은 당연하다 생각 할 수 있지만 타인은 다르게 생각 할 수 있습니다.
대표적으로 dir (directory, direct), res(result, resource)

정말 모두가 쓰는 약어만을 사용하시는게 좋습니다.
HTML, ID (PW는 아닙니다),  idx, cnt, asc, desc, num, arr, str 같은 것들이 있습니다.
위에서 언급한 dir 도 사용처가 다르기 때문에 구분 할 수 있다는 의견도 있네요.

저는 개인적으로 idx, cnt 같은 약어도 안쓰려고 합니다.
이건 개인 성향 차이가 있을 수 있겠네요.

comforest reviewed (changes requested):
간단히 네이밍 / 컴포넌트 구성에 대해서만 코멘트 달아 드렸어요.
\<div\> 사용이 많으신데 [\<Fragment\>](https://ko.reactjs.org/docs/fragments.html) 도 공부하시면 좋을 것 같아요.

추후에 디렉토리 구조 같은것도 이야기 해봐요.

---
커밋 메시지는 제가 따로 이야기 해드렸지만 다시 언급하겠습니다
커밋은 작업 단위로 나누어서 커밋 해주셔야 해요.
어떤 화면을 추가했는지, 기능(버튼 클릭 등)을 추가했는지, 디자인을 바꿨다는지
또한 그 작업 내용을 커밋 메시지에 자세히 작성 부탁 드릴게요

tankboy091e commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+        setValue(event.target.value);
+    };
+    
+    const sendFunc = () => {
+        if(value === '')  return;
+        let tok = false;
+        let idx = 0;
+        if(Math.floor(Math.random() * 2)){
+            tok = true;
+            idx = Math.floor(Math.random() * userText.length);
+        }
+        if(tok){
+            setMessage([...message, {'isMine': true, 'text': value}, {'isMine': false, 'text': userText[idx]}]);
+        }else{
+            setMessage([...message, {'isMine': true, 'text': value}]);
+        }
+        setValue('');
+    }
```
```suggestion
    const sendMessage = () => {
        if(textarea.current.value === '') return;
        const random = Math.random() < 0.5
        const myMessage = {isMine : true, text : value}
        const yourMessage = userText[~~(random * userText.length)]
        setMessage([...message, myMessage].concat(random === true ? yourMessage : []))
    }
```

이렇게도 작성 가능합니다.

tankboy091e commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+                {name}
+            </div>
+            <div className="chatBox">
+                {message.map(item => {
+                    return item['isMine'] ? <div style={{textAlign: 'right'}}><div className="mySendBox">{item['text']}</div></div> : <div className="userSendBox">{item['text']}</div>
+                })}
```
```suggestion
                {message.map(({isMine, text})=> {
                  return <div className={`${isMine === true 
                  ? styles.mySendBox  
                  : styles.yourSendBox}`}>{text}</div> })}
```
1. 삼항 연산자는 가장 작은 단위에서 써야 코드의 중복을 피할 수 있습니다.

2. map의 value 객체에 담긴 여러 필드에 참조를 반복해야 한다면 분해할당을 이용해 보세요.

3. mySendBox의 textAlign을 조정하기 위해 새로운 div를 사용하셨습니다. 단순히 클래스네임을 바꾸는 것으로 해결할 수 있습니다.

blackSooooo replied:
분해할당에 대해 찾아보고 알게 됐어요. 감사합니다:) 그런데 textAlign을 조정하려면 새로운 div가 필요한게 아닌가요? 그래야 새로운 div 안에 있는 것들이 오른쪽으로 정렬되는게 아닌가요?

tankboy091e replied:
딱 정해놓고 말씀드릴 수 없는 것은 방법이 여러가지가 있기 때문입니다. 또 컴포넌트 내부에 어떤 요소가 들어가는지에 따라 다르겠지요. 좌우를 조정할 때에 현수님께서 제시한 방법도 유효합니다. 그러나 차라리 말씀드리고 싶었던 바는 매우 유사한 역할을 하는 동일한 두 객체는 그 구조도 유사해야 한다는 것입니다. textAlign을 바꾸기 위해 div를 하나 더 쓴다면, 너비와 높이, 간격 혹은 정렬에서 문제가 발생할 수 있습니다. 여기서 문제는 종종 동일한 결과가 산출이 되지 않는 것과, 유지보수에 있어서 번거로워 진다는 것입니다.

tankboy091e commented on src/pages/ChatPage.css:
```diff
@@ -0,0 +1,78 @@
+.container {
+    border-width: 10px;
+    height: 100vh;
+    width: 30vw;
```
최고 부모 위치에 html, body, 예컨대 layout 등의 element가 아니라면 vw와 vh를 사용하시는 것은 위험합니다. %를 사용하셔야 합니다.

tankboy091e commented on src/pages/ChatPage.css:
```diff
@@ -0,0 +1,78 @@
+    background-color: #fce44d;
+    border: none;
+    color:#000000;
+    padding: 15px 0;
+    border-radius: 8px;
+    font-size: 15px;
```
padding이나 margin, font-size를 px로 설정하시는 것은 특수한 경우에 따라서입니다. 보통은 rem이나 em을 사용해야 한다고 여겨집니다. 뷰에 따라 유기적으로 글자 크기와 간격을 조정하기 위해서입니다.

tankboy091e reviewed (changes requested):
css module을 사용하는 방식과, useRef, 또 삼항 연산자를 이용해서 jsx element를 return하는 방식, 그리고 몇 가지 나머지 것들에 대해 말씀드렸습니다.

MainPage에서 받은 전역적 데이터를 다른 라우트의 페이지에서 이용해보시는 것도 도움이 되시리라 생각합니다.

comforest commented on src/pages/ChatPage.js:
```diff
@@ -0,0 +1,46 @@
+    const [message, setMessage] = useState([]);
+    const userText=['안녕하세요', '샘플 데이터입니다', '반가워요', '날씨가 어때요?', '감사합니다', '좋은하루 되세요', '사랑해요ㅎ.ㅎ'];
+    const [value, setValue] = useState('');
+    const onChange = (event) => {
+        setValue(event.target.value);
+    };
```
React 공식 문서에 value 에 state를 써서 쓰는 코드가 있어서 저도 아무 생각 없이 사용했었네요.

https://www.codebeast.dev/usestate-vs-useref-re-render-or-not/

같이 읽어보면 좋을 것 같아요

blackSooooo replied:
기존에 구현한 것은 텍스트를 입력하면 작성버튼 스타일이 다르게 적용되게 구현을 했는데 
```
 <input type="submit" className={`${inputMessage.length === 0 ? styles.sendBox : styles.sendTextingBox}`} value="전송"/>
```

useRef를 이용하면 onSubmit을 하기 전까지 value 변경이 반영이 안되는데 혹시 이를 해결할 방법이 있을까요?

또, 위에서 언급해주신대로 form을 이용해서 엔터를 누르면 처리가 되게끔 구현을 해보았는데 
바로 화면에서 잠깐 반영됐다가 전부 다 사라지는 이유가 무엇일까요,,,

```
const onChangeInputMessage = e => {
        setInputMessage(e.target.value);
};
const sendMessage = () => {
        if(inputMessage === '')  return;
        Math.random() < 0.5 
            ? setMessages([...messages, {'isMine': true, 'text': inputMessage}, {'isMine': false, 'text': userTexts[Math.floor(Math.random() * userTexts.length)]}])
            : setMessages([...messages, {'isMine': true, 'text': inputMessage}])
        setInputMessage('');
}


<div className={styles.inputBox}>
        <form onSubmit={sendMessage}>
            <input className={styles.textBox} onChange={onChangeInputMessage}/>
            <input type="submit" className={`${inputMessage.length === 0 ? styles.sendBox : styles.sendTextingBox}`} value="전송"/>
        </form>
</div>
```

tankboy091e replied:
첫 번째 경우를 해결하기 위해서는 useState를 쓰시는 것이 맞습니다. 현수님께서 원하시는 것이 텍스트 필드에 값이 있는지의 여부라면 onChange 이벤트에서 매번 state를 갱신할 필요는 없겠지요. onChange에 할당한 함수 내부에서 inputMessage의 value가 비어있는지 아닌지만 검사해서 state를 갱신해주세요. 가령 다음과 같을 것입니다.

```suggestion
    const [active, setActive] = useState(false)
    
    const onInputMessageChange = (event) => {
        setActive(event.target.value !== '')
    };
    
    return <!--생략--> <input className={`${styles.submit} ${active === true
        : styles.active
        : styles.inactive }`} type='submit' />
```

setState는 기왕의 값과 동일하다면 자체적으로 컴포넌트의 갱신을 수행하지 않으니 별도로 분기처리할 필요는 없습니다.
submit의 클래스네임을 처리한 것에 주목해주세요. css를 완전히 별도로 작성하지 않고, 예컨대 변경되어야 할 색깔과 같은 부분만 .inactive, .active에 구분해서 작성하실 수 있습니다.

formEvent는 http request를 보냅니다. 즉, 페이지가 자동적으로 갱신되는데요, 폼이벤트 전송 후 라우트가 변경되는 페이지면 상관이 없겠지만 리액트와 같은 SPA에서는 문제가 됩니다. formEvent가 주입하는 event 파라미터에서 다음의 메서드를 호출하세요(더 자세하게는 검색해 보실 수 있습니다).

```suggestion
    e.preventDefault()
```

blackSooooo committed:
feat: 입력 변수명 변경, onSubmit추가

blackSooooo committed:
[feat] ChatPage css파일 수정
1. module.css로 파일 내용 변경
2. rem, em을 활용

blackSooooo committed:
[feat] ChatPage 기능 수정
1. 화면 크기 조정, 뒤로가기 추가
2. setState시 리랜더링 되는 것을 방지하고자 useRef를 이용해서 input 설정
3. 변수명 변경 및 분해할당 사용

comforest commented on src/pages/ChatPage.js:
```diff
@@ -1,44 +1,57 @@
-        setValue(event.target.value);
+    const [messages, setMessages] = useState([]);
+    const userTexts=['안녕하세요', '샘플 데이터입니다', '반가워요', '날씨가 어때요?', '감사합니다', '좋은하루 되세요', '사랑해요ㅎ.ㅎ'];
+    const [typing, setTyping] = useState(false);
+    const [fullScreen, setFullScreen] = useState(false);
```
```suggestion
    const [isTyping, setIsTyping] = useState(false);
    const [isFullScreen, setIsFullScreen] = useState(false);
```

boolean 형태 변수명은 is~~ 형태를 많이 사용합니다.

comforest commented on src/pages/ChatPage.js:
```diff
@@ -1,44 +1,57 @@
+    const userTexts=['안녕하세요', '샘플 데이터입니다', '반가워요', '날씨가 어때요?', '감사합니다', '좋은하루 되세요', '사랑해요ㅎ.ㅎ'];
+    const [typing, setTyping] = useState(false);
+    const [fullScreen, setFullScreen] = useState(false);
+    const inputRef = useRef();
```
```suggestion
    const messageInputRef = useRef();
```

추후 input이 더 추가될 수 도 있으니 어떠한 input인지 조금 더 명확히 할 필요가 있습니다.

comforest commented on src/pages/ChatPage.js:
```diff
@@ -1,44 +1,57 @@
+        e.preventDefault();
+        if(inputRef.current.value === '')  return;
+        Math.random() < 0.5 
+            ? setMessages([...messages, {'isMine': true, 'text': inputRef.current.value}, {'isMine': false, 'text': userTexts[Math.floor(Math.random() * userTexts.length)]}])
+            : setMessages([...messages, {'isMine': true, 'text': inputRef.current.value}])
```
나중에 코드가 길어지면 2번째 인자 { isMine, text } 는 앞에서 따로 변수로 만들어두는게 더 편할거에요.

blackSooooo commented:
말씀해주신 부분들 반영해서 수정해봤습니다.

comforest commented on src/pages/ChatPage.js:
```diff
@@ -1,44 +1,57 @@
+                </div>
+                <div className={styles.chatBox}>
+                    {messages.map(({isMine, text}) => {
+                        return <div style={isMine===true ? {textAlign: 'right'} : {}}><div className={`${styles.sendBox} ${isMine === true
+                        ? styles.myMessage : styles.userMessage}`}>{text}</div></div>})}
```
저번에도 진수님께서 여기 조언 해주신 것 같은데
아직까지 가독성이 많이 떨어집니다 ㅠㅜㅠ
이번엔 여기까지 하시고
나중에 ESLint 같은거 쓰면 빨간줄 매우 많이 뜰것 같네요 ㅎㅎㅎ...

comforest commented on src/pages/MainPage.js:
```diff
@@ -2,18 +2,20 @@ import React, { useState } from 'react';
-    const onChange = (event) => {
-        setValue(event.target.value);
+    const [name, setName] = useState('');
+    const onChangeName = (e) => {
+      setName(e.target.value);
     }
```
onChange 랑 onSubmit 동시에 이 함수를 사용하는게 조금 신기하네요...
일단 함수명은 onChange에서만 쓸 것 처럼 써져 있어서 헷갈리게 되어있네요

동작은 잘 되는 것 처럼 보이지만 e.preventDefault() 가 없어 페이지 전체 갱신이 되고 있을테고...
흠 시험 기간 끝나고 좀 더 자세히 이야기 해봐요

comforest reviewed (changes requested):
일단 커밋 메시지가 매우 좋아졌더라구요 ㅎㅎㅎ
[feat] 대신 "feat: " 으로 해주시면 될 것 같아요.
그리고 수정은 fix 에 가까우니
"Chatpage 기능 수정" 보단 "ChatPage 기능 추가" 가 더 적절 해보이기도 하구요

"Chatpage css 파일 수정" 쪽은 기능 추가 아니기 때문에 fix 접두사가 더 적절해 보입니다.
그리고 해당 커밋에 import 수정 사항이 반영이 안되어 있어요. (ChatPage.js:2 - 두번째 줄이란뜻) 이 커밋으로 롤백(정확힌 체크아웃 이란 표현을 씁니다)을 하게되면 에러가 발생하겠죠??

tankboy091e commented on src/pages/ChatPage.js:
```diff
@@ -1,44 +1,57 @@
+        <div style={{backgroundColor: 'rgb(225,225,225)'}}>
+            <div className={`${styles.screen} ${fullScreen ? styles.big : styles.small}`}>
+                <div className={styles.header}>
+                    <div style={{paddingTop: 10}}> 
```
styled component에 비해서 css module이 갖는 작은 단점은 한 눈에 알아보기가 약간 어렵다는 건데요, 예컨대 styled component를 사용하면 `<Container><Container>`와 같이 적을 수 있는 대목을 `<div className={styles.container}></div>`처럼 적어야 한다는 것입니다. 단순히 코드의 가독성뿐만 아니라, 역할을 부여하는 데에 있어서 어려움을 겪으실 수 있습니다. `<div>`는 CSS 스타일을 임기응변하기 위해서 새로 놓게 되면 보시는 코드와 같이 지나치게 깊고 장황해집니다. style을 위해서 마크업을 사용하시는 것보다는, 분명한 역할을 생각해두셔야 합니다. 예를 들어 이것은 Container다, Label이다, 혹은 별 다섯 개를 포함하는 평점이다 하는 것과 같이 말입니다. Tricky한 애니메이션을 위해서 어쩔 수 없이 div로 감싸야 하는 특별한 경우를 제외하면 최대한 간결해지려고 하시는 것이 중요하다 생각됩니다.

tankboy091e reviewed (commented):
늦게 리뷰를 드려서 죄송합니다. 다양한 레이아웃을 최소한의 jsx element로 구현해보시면 많은 도움이 될 것 같습니다.
